internal/registry/registry_cache.go

```diff
@@ -63,54 +62,73 @@ func NewRegistryCache(logger logging.Logger, home, registryURL string) (Cache, e
 	}, nil
 }
 
-func NewDefaultRegistryCache(logger logging.Logger, home string) (Cache, error) {
-	return NewRegistryCache(logger, home, defaultRegistryURL)
-}
-
-func (r *Cache) createCache() error {
-	r.logger.Debugf("Creating registry cache for %s/%s", r.url.Host, r.url.Path)
-
-	root, err := ioutil.TempDir("", "registry")
+// LocateBuildpack stored in registry
+func (r *Cache) LocateBuildpack(bp string) (Buildpack, error) {
+	err := r.Refresh()
 	if err != nil {
-		return err
+		return Buildpack{}, errors.Wrap(err, "refreshing cache")
 	}
 
-	repository, err := git.PlainClone(root, false, &git.CloneOptions{
-		URL: r.url.String(),
-	})
+	ns, name, version, err := buildpack.ParseRegistryID(bp)
 	if err != nil {
-		return errors.Wrap(err, "could not clone remote registry")
+		return Buildpack{}, errors.Wrap(err, "parsing buildpacks registry id")
 	}
 
-	w, err := repository.Worktree()
+	entry, err := r.readEntry(ns, name)
 	if err != nil {
-		return err
+		return Buildpack{}, errors.Wrap(err, "reading entry")
 	}
 
-	return os.Rename(w.Filesystem.Root(), r.Root)
+	if len(entry.Buildpacks) > 0 {
+		if version == "" {
+			highestVersion := entry.Buildpacks[0]
+			if len(entry.Buildpacks) > 1 {
+				for _, bp := range entry.Buildpacks[1:] {
+					if semver.Compare(fmt.Sprintf("v%s", bp.Version), fmt.Sprintf("v%s", highestVersion.Version)) > 0 {
+						highestVersion = bp
+					}
+				}
+			}
+			return highestVersion, highestVersion.Validate()
+		}
```

Comment:
This gets pretty deeply nested, so I would be tempted to move this in to a few helper methods or a service.

Could `Entry` provide us a `BuildpackForVersion(version string)`?

I feel like this would only return information about the data structure rather than perform another service, so I treat is as a single concern. If we wanted to strongly maintain a separation of concerns could we add a `RegistryCacheManager` or something similar to provide this method?

Within that method could we use some private helpers such as `highestVersion()`? Could we use tail recursion to iterate through our list rather than a range?

Reply:
I definitely hear that. I was trying to avoid a larger scale refactor in this PR, but I can definitely move this into a separate func (perhaps not on `Entry`, so as to keep it off data structures, though perhaps I should also move validation off `Buildpacks`...)